### ISSUE TITLE
Potential API breakage: Base36qp

### ISSUE THREAD

**Issue by lukateras** [lukateras/rust-base36#1]
@zo-el (Holo), @mplanchard (CUIDv1), @ppodolsky (Summa):

Sorry to bother you all, but this is kind of important.

`base36` [v0.0.1](https://crates.io/crates/base36/0.0.1) uses the following Base36 alphabet:

https://github.com/lukateras/rust-base36/blob/4e6981e13ec5f3748dfe9b8870aa4e8ed749096d/src/lib.rs#L3

...Which is not standard. `q` and `p` are swapped places. My typo. (I'll call this alternate encoding Base36qp from now on.)

As a result, `Hello, World!` encoded to Base36qp is `fg3h7vpw7een6jwwnzmq`, while standard Base36 would be `fg3h7vqw7een6jwwnzmp`.

I'm assuming that some of you have baked this alternate Base36, Base36qp, into the project spec. Is that the case, or would it be easy for you to swap out the encoding for standard Base36?

I've since reimplemented `base36` on top of the `num` crate, written documentation, and expanded the test suite. I'm planning to publish v1.0.0 as soon as I ensure that I don't cause too many problems for any of you.

If anyone does indeed depend on Base36qp, here are the options I'm considering to smooth out the upgrade path:

- documenting how to swap `q`s and `p`s after `encode` (before `decode`)
- adding functions `encode_qp` and `decode_qp` that are compatible with `base36` v0.0.1 `encode` and `decode`
- publishing another crate, `base36qp`, similar to this one but preserving compatibility with `base36` v0.0.1

If you have a preference, let me know.

**Comment by mplanchard** [comment]
Thanks for the ping!

I believe that in CUID's case, the inclusion of `base36` in the `Cargo.toml` file was actually vestigial. We moved a while back to a performance-optimized custom implementation. In addition, at least in our case, base36 encoding was used as part of transient ID generation just to represent random data in a more compact fashion: there was no expectation of the IDs ever being reversible, so the mismatch would not have harmed CUID generation.

As such, at least for us, there's no need to maintain backwards compatibility with the qp version or to provide methods for it.

**Comment by lukateras** [comment]
@mplanchard: I'm glad!

**Comment by lukateras** [comment]
@zo-el has privately told me they're fine with the breaking change to the standard Base36 as well.

**Comment by lukateras** [comment]
@ppodolsky @JarvisCraft @vinogradovkonst (Summa):

Happy holidays! (Sorry to bother you all right before the New Year!)

I'm planning to publish the update that switches to standard Base36 by mid-January at the latest. The release tags have already been cut: [`v1.0.0`](https://github.com/lukateras/rust-base36/releases/tag/v1.0.0) and [`v1.0.1`](https://github.com/lukateras/rust-base36/releases/tag/v1.0.1).

In short, with this update the encoded `q`s will become `p`s, and vice versa. I've appropriately bumped the major version, but still, I'd like to avoid any unpleasant surprises on your side.

Could you check if Summa maintains compatibility if updated to the newer version of the `base36` crate? I'd be happy to help you migrate (:

(Кстати, можно и по-русски!)

**Comment by lukateras** [comment]
I'm going to publish the update as-is. If anybody needs any help, just let me know (:
